slattach/slattach.8
.\" Copyright (c) 1986 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted
.\" provided that the above copyright notice and this paragraph are
.\" duplicated in all such forms and that any documentation,
.\" advertising materials, and other materials related to such
.\" distribution and use acknowledge that the software was developed
.\" by the University of California, Berkeley.  The name of the
.\" University may not be used to endorse or promote products derived
.\" from this software without specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
.\" WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)slattach.8	6.2 (Berkeley) 9/19/88
.\"
.TH SLATTACH 8 "September 19, 1988"
.UC 6
.SH NAME
slattach \- attach serial lines as network interfaces
.SH SYOPNSIS
.B /etc/slattach
ttyname [
.I baudrate
]
.SH DESCRIPTION
.I Slattach
is used to assign a tty line to a network interface,
and to define the network source and destination addresses. The
.I ttyname
parameter is a string of the form ``ttyXX'', or ``/dev/ttyXX''.
The optional
.I baudrate
parameter is used to set the speed of the connection. If not specified, the
default of 9600 is used.
.PP
Only the super-user may attach a network interface.
.PP
To detach the interface, use `ifconfig
.I interface-name
down' after killing off the
.I slattach
process.
.I interface-name
is the name that is shown by
.B netstat(1)
.SH EXAMPLES
.ta 8
	/etc/slattach ttyh8
.br
	/etc/slattach /dev/tty01 4800
.SH DIAGNOSTICS
Messages indicating the specified interface does not exit, the
requested address is unknown, the user is not privileged and
tried to alter an interface's configuration.
.SH "SEE ALSO"
netstat(1), netintro(4), ifconfig(8), rc(8)


named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


usr/src/sbin/slattach/slattach.c
#include <stdio.h>
#include <sys/types.h>
#include <sgtty.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <net/if.h>
#include <netdb.h>
#include <fcntl.h>

#ifndef lint
static char rcsid[] = "$Header: slattach.c,v 1.1 84/10/04 12:57:12 rick Exp $";
#endif

struct	ifreq ifr;
struct	sockaddr_in sin = {AF_INET};

#define DEFAULT_BAUD	9600
int	speed;
int	slipdisc = SLIPDISC;

char	devname[32];
char	hostname[32];

extern int errno;

main(argc, argv)
	int argc;
	char *argv[];
{
	register FILE *fp;
	register int fd;
	register char *dev = argv[1];
	struct sgttyb sgtty;
	int n;

	if (argc < 4 || argc > 5)
		error(1, 0, "usage: %s ttyname srcaddr dstaddr [baud]\n",
		      argv[0]);
	speed = argc == 5 ? findspeed(atoi(argv[4])) : findspeed(DEFAULT_BAUD);
	if (speed == 0)
		error(1, errno, "unknown speed %s", argv[4]);
	if (strncmp("/dev/", dev, 5)) {
		sprintf(devname, "/dev/%s", dev);
		dev = devname;
	}
	if ((fd = open(dev, O_RDWR | O_NDELAY)) < 0)
		error(1, errno, "%s: cannot open", dev);
	sgtty.sg_flags = RAW | ANYP;
	sgtty.sg_ispeed = sgtty.sg_ospeed = speed;
	if (ioctl(fd, TIOCSETP, &sgtty) < 0)
		error(1, errno, "ioctl(TIOCSETP) on %s", dev);
	if (ioctl(fd, TIOCSETD, &slipdisc) < 0)
		error(1, errno, "ioctl(TIOCSETD)");
	if (ioctl(fd, TIOCGETD, &n) < 0)
		error(1, errno, "ioctl(TIOCGETD)");

	fd = socket(AF_INET, SOCK_DGRAM, 0);
	if (fd < 0)
		error(1, errno, "socket(AF_INET, SOCK_DGRAM)");
	sprintf(ifr.ifr_name, "sl%d", n);

#ifdef sun
	/*
	 * This crap is necessary because sun changed the ioctl handling so
	 * that SIOCSIFDSTADDR ONLY gets handed the value of ifr_data, not
	 * the whole ifr structure like any sane implementation would. This
	 * is clearly a mistake on their part. Fortunately, sizeof (caddr_t)
	 * == sizeof(sin.sin_addr.s_addr) if it's not on your machine, you
	 * lose. 
	 */
	getaddr(argv[3], &sin);
	ifr.ifr_data = (caddr_t) sin.sin_addr.s_addr;
#else vax
	getaddr(argv[3], (struct sockaddr_in *)&ifr.ifr_dstaddr);
#endif vax
	if (ioctl(fd, SIOCSIFDSTADDR, (caddr_t)&ifr) < 0)
		error(1, errno, "ioctl(SIOCSIFDSTADDR)");
	getaddr(argv[2], (struct sockaddr_in *)&ifr.ifr_addr);
	if (ioctl(fd, SIOCSIFADDR, (caddr_t)&ifr) < 0)
		error(1, errno, "ioctl(SIOCSIFADDR)");
	for (;;)
		sigpause(0);
}

struct sg_spds {
	int sp_val, sp_name;
}       spds[] = {
#ifdef B50
	{ 50, B50 },
#endif
#ifdef B75
	{ 75, B75 },
#endif
#ifdef B110
	{ 110, B110 },
#endif
#ifdef B150
	{ 150, B150 },
#endif
#ifdef B200
	{ 200, B200 },
#endif
#ifdef B300
	{ 300, B300 },
#endif
#ifdef B600
	{ 600, B600 },
#endif
#ifdef B1200
	{ 1200, B1200 },
#endif
#ifdef B1800
	{ 1800, B1800 },
#endif
#ifdef B2000
	{ 2000, B2000 },
#endif
#ifdef B2400
	{ 2400, B2400 },
#endif
#ifdef B3600
	{ 3600, B3600 },
#endif
#ifdef B4800
	{ 4800, B4800 },
#endif
#ifdef B7200
	{ 7200, B7200 },
#endif
#ifdef B9600
	{ 9600, B9600 },
#endif
#ifdef EXTA
	{ 19200, EXTA },
#endif
#ifdef EXTB
	{ 38400, EXTB },
#endif
	{ 0, 0 }
};

findspeed(speed)
	register int speed;
{
	register struct sg_spds *sp;

	sp = spds;
	while (sp->sp_val && sp->sp_val != speed)
		sp++;
	return (sp->sp_name);
}

struct in_addr inet_makeaddr();

getaddr(s, sin)
	char *s;
	register struct sockaddr_in *sin;
{
	register struct hostent *hp;
	register struct netent *np;
	int val;

	bzero(sin, sizeof(struct sockaddr_in));
	hp = gethostbyname(s);
	if (hp) {
		sin->sin_family = hp->h_addrtype;
		bcopy(hp->h_addr, (char *) &sin->sin_addr, hp->h_length);
		return;
	}
	np = getnetbyname(s);
	if (np) {
		sin->sin_family = np->n_addrtype;
		sin->sin_addr = inet_makeaddr(np->n_net, INADDR_ANY);
		return;
	}
	sin->sin_family = AF_INET;
	val = inet_addr(s);
	if (val != -1) {
		sin->sin_addr.s_addr = val;
		return;
	}
	val = inet_network(s);
	if (val != -1) {
		sin->sin_addr = inet_makeaddr(val, INADDR_ANY);
		return;
	}
	error(1, 0, "%s: bad value\n", s);
}
